File: src/main/java/no/nav/jfr/generell/operations/eventenricher/EventEnricher.java
package no.nav.jfr.generell.operations.eventenricher;

import no.nav.jfr.generell.infrastructure.exceptions.*;
import no.nav.jfr.generell.infrastructure.metrics.Metrics;
import no.nav.jfr.generell.operations.SkjemaMetadata;
import no.nav.jfr.generell.operations.eventenricher.journalpost.Journalpost;
import no.nav.jfr.generell.operations.eventenricher.pdl.Ident;
import no.nav.jfr.generell.operations.eventenricher.pdl.PdlClient;
import no.nav.jfr.generell.infrastructure.kafka.EnrichedKafkaEvent;
import no.nav.jfr.generell.infrastructure.kafka.KafkaEvent;
import no.nav.jfr.generell.operations.eventenricher.saf.SafClient;
import no.nav.jfr.generell.operations.generell.felleskodeverk.FkvClient;
import no.nav.jfr.generell.operations.generell.felleskodeverk.FkvKrutkoder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Arrays;
import java.util.List;
import java.util.TimerTask;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class EventEnricher {
    private static final Logger log = LoggerFactory.getLogger(EventEnricher.class);
    private static final long KRUT_KODER_REFRESH_INTERVAL_HOURS = 5L;
    private final SafClient safClient;
    private final PdlClient pdlClient;
    private final FkvClient fkvClient;
    private FkvKrutkoder fkvKrutkoder;
    private SkjemaMetadata skjemaMetadata;

    public EventEnricher() throws Exception {
        this.safClient = new SafClient();
        this.pdlClient = new PdlClient();
        this.fkvClient = new FkvClient();
        this.skjemaMetadata = new SkjemaMetadata();
        this.fkvKrutkoder = fkvClient.fetchKrutKoder();
        this.setupRefreshFellesKodeverkTimer();
    }

    public void createEnrichedKafkaEvent(final EnrichedKafkaEvent enrichedKafkaEvent) throws Exception{
        enrichJournalpostInKafkaEvent(enrichedKafkaEvent);
        updateBehandlingValues(enrichedKafkaEvent);
        if(enrichedKafkaEvent.isPersonbruker()){
            enrichIdenterFromPDL(enrichedKafkaEvent);
        }else{
            log.info("Kaller ikke PDL da bruker på journalpost {} er ikke en person", enrichedKafkaEvent.getJournalpostId());
        }
    }

    private void enrichJournalpostInKafkaEvent(final EnrichedKafkaEvent enrichedKafkaEvent)throws Exception{
        final Journalpost journalpost = retriveJournalpostFromSaf(enrichedKafkaEvent.getKafkaEvent());
        enrichedKafkaEvent.setJournalpost(journalpost);
        if(journalpost.invalidJournalpostStatus()){
            log.info("Avslutter videre behandling da journalpost {} har status {}", journalpost.getJournalpostId(), journalpost.getJournalstatus());
            Metrics.incInvalidJournalpostStatus(enrichedKafkaEvent);
            throw new InvalidJournalpostStatusException();
        }
        if(skjemaMetadata.isIgnoreskjema(enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getSkjema())){
            log.info("Avslutter videre behandling da journalpost {} har skjema {} på tema {} som eksplisitt skal ignoreres!", enrichedKafkaEvent.getJournalpostId(), enrichedKafkaEvent.getSkjema(), enrichedKafkaEvent.getTema());
            throw new InvalidJournalpostStatusException();
        }
        enrichedKafkaEvent.setJournalpost(journalpost);
    }

    private Journalpost retriveJournalpostFromSaf(final KafkaEvent kafkaEvent) throws TemporarilyUnavailableException, ExternalServiceException {
        try {
            final Journalpost journalpost = safClient.retriveJournalpost(kafkaEvent.getJournalpostId());
            log.info("Hentet {} fra Saf", journalpost.toString());
            return journalpost;
        } catch (final ExternalServiceException se) {
            log.error("Kunne ikke hente journalpost: {} fra Saf" , kafkaEvent.getJournalpostId());
            throw se;
        }
    }

    private void enrichIdenterFromPDL(EnrichedKafkaEvent enrichedKafkaEvent) throws Exception {
        Journalpost journalpost = enrichedKafkaEvent.getJournalpost();
        if(journalpost.getBruker() == null){
            log.info("Bruker er ikke satt på journalpost: {}. Kan ikke hente fra PDL", journalpost.getJournalpostId());
            throw new FunctionalRequirementException();
        }
        final List<Ident> identer = pdlClient.retrieveIdenterFromPDL(journalpost.getBruker().getId(), journalpost.getTema(), enrichedKafkaEvent.getJournalpostId());
        enrichedKafkaEvent.setIdenter(identer);
        log.info("Hentet alle id-nummer for bruker på journalpost: {} fra PDL", journalpost.getJournalpostId());
    }

    private void updateBehandlingValues(final EnrichedKafkaEvent enrichedKafkaEvent) throws FunctionalRequirementException {
        final Journalpost journalpost = enrichedKafkaEvent.getJournalpost();
        try {
            String behandlingsTema = fkvKrutkoder.getBehandlingstema(journalpost.getTema(), journalpost.getBrevkode());
            String behandlingsType = fkvKrutkoder.getBehandlingstype(journalpost.getTema(), journalpost.getBrevkode());
            if (!(journalpost.getBehandlingstema() == null || journalpost.getBehandlingstema().isEmpty())) {
                behandlingsTema = journalpost.getBehandlingstema();
            }
            journalpost.setBehandlingstema(behandlingsTema);
            journalpost.setBehandlingstype(behandlingsType);
            log.info("Satt følgende verdier behandlingstema: '{}' og behandlingstype: '{}' på journalpost {}", behandlingsTema, behandlingsType, journalpost.getJournalpostId());
        } catch (Exception e) {
            enrichedKafkaEvent.setToFordeling(true);
            enrichedKafkaEvent.setToManuell(true);
            log.warn("Klarte ikke finne behandlingsverdier for journalpost {} med tema {} og skjema {}", journalpost.getJournalpostId(), journalpost.getTema(), journalpost.getBrevkode(), e);
        }
    }

    private void setupRefreshFellesKodeverkTimer() {
        final TimerTask fellesKodeverkRefresher = new TimerTask() {
            @Override
            public void run() {
                try {
                    fkvKrutkoder = fkvClient.fetchKrutKoder();
                } catch (Exception e) {
                    log.error("Kunne ikke oppdatere KrutKoder fra FKV. Bruker gamle koder og prøver på nytt senere");
                }
            }
        };
        final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
        long refresherDelay = KRUT_KODER_REFRESH_INTERVAL_HOURS;
        executor.scheduleAtFixedRate(fellesKodeverkRefresher, refresherDelay, refresherDelay, TimeUnit.HOURS);
    }
}

File: src/main/kotlin/no/nav/helse/flex/operations/eventenricher/EventEnricher.kt
package no.nav.helse.flex.operations.eventenricher

import no.nav.helse.flex.infrastructure.exceptions.ExternalServiceException
import no.nav.helse.flex.infrastructure.exceptions.FunctionalRequirementException
import no.nav.helse.flex.infrastructure.exceptions.InvalidJournalpostStatusException
import no.nav.helse.flex.infrastructure.kafka.EnrichedKafkaEvent
import no.nav.helse.flex.infrastructure.kafka.KafkaEvent
import no.nav.helse.flex.infrastructure.metrics.Metrics.incInvalidJournalpostStatus
import no.nav.helse.flex.operations.SkjemaMetadata
import no.nav.helse.flex.operations.eventenricher.journalpost.Journalpost
import no.nav.helse.flex.operations.eventenricher.pdl.PdlClient
import no.nav.helse.flex.operations.eventenricher.saf.SafClient
import no.nav.helse.flex.operations.generell.felleskodeverk.FkvClient
import no.nav.helse.flex.operations.generell.felleskodeverk.FkvKrutkoder
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class EventEnricher(
    private val safClient: SafClient = SafClient(),
    private val pdlClient: PdlClient = PdlClient(),
    private val fkvClient: FkvClient = FkvClient(),
    private val skjemaMetadata: SkjemaMetadata = SkjemaMetadata()
) {
    private val log = LoggerFactory.getLogger(EventEnricher::class.java)
    private val KRUT_KODER_REFRESH_INTERVAL_HOURS = 5L

    private var fkvKrutkoder: FkvKrutkoder

    init {
        fkvKrutkoder = fkvClient.fetchKrutKoder()
        setupRefreshFellesKodeverkTimer()
    }

    fun createEnrichedKafkaEvent(enrichedKafkaEvent: EnrichedKafkaEvent) {
        enrichJournalpostInKafkaEvent(enrichedKafkaEvent)
        updateBehandlingValues(enrichedKafkaEvent)
        if (enrichedKafkaEvent.isPersonbruker) {
            enrichIdenterFromPDL(enrichedKafkaEvent)
        } else {
            log.info("Kaller ikke PDL da bruker på journalpost ${enrichedKafkaEvent.journalpostId} er ikke en person")
        }
    }

    private fun enrichJournalpostInKafkaEvent(enrichedKafkaEvent: EnrichedKafkaEvent) {
        val journalpost = retriveJournalpostFromSaf(enrichedKafkaEvent.kafkaEvent)
        enrichedKafkaEvent.journalpost = journalpost
        if (journalpost!!.invalidJournalpostStatus()) {
            log.info("Avslutter videre behandling da journalpost ${journalpost.journalpostId} har status ${journalpost.journalstatus}")
            incInvalidJournalpostStatus(enrichedKafkaEvent)
            throw InvalidJournalpostStatusException()
        }
        if (skjemaMetadata.isIgnoreskjema(enrichedKafkaEvent.tema, enrichedKafkaEvent.skjema)) {
            log.info("Avslutter videre behandling da journalpost ${enrichedKafkaEvent.journalpostId} har skjema ${enrichedKafkaEvent.skjema} på tema ${enrichedKafkaEvent.tema} som eksplisitt skal ignoreres!")
            throw InvalidJournalpostStatusException()
        }
        enrichedKafkaEvent.journalpost = journalpost
    }

    private fun retriveJournalpostFromSaf(kafkaEvent: KafkaEvent): Journalpost? {
        return try {
            val journalpost = safClient.retriveJournalpost(kafkaEvent.getJournalpostId())
            log.info("Hentet $journalpost fra Saf")
            journalpost
        } catch (se: ExternalServiceException) {
            log.error("Kunne ikke hente journalpost: ${kafkaEvent.getJournalpostId()} fra Saf")
            throw se
        }
    }

    private fun enrichIdenterFromPDL(enrichedKafkaEvent: EnrichedKafkaEvent) {
        val journalpost = enrichedKafkaEvent.journalpost!!
        if (journalpost.bruker == null) {
            log.info("Bruker er ikke satt på journalpost: ${journalpost.journalpostId}. Kan ikke hente fra PDL")
            throw FunctionalRequirementException()
        }

        val identer = pdlClient.retrieveIdenterFromPDL(
            journalpost.bruker!!.id,
            journalpost.tema,
            enrichedKafkaEvent.journalpostId
        )
        enrichedKafkaEvent.identer = identer
        log.info("Hentet alle id-nummer for bruker på journalpost: ${journalpost.journalpostId} fra PDL")
    }

    private fun updateBehandlingValues(enrichedKafkaEvent: EnrichedKafkaEvent) {
        val journalpost = enrichedKafkaEvent.journalpost!!

        try {
            var behandlingsTema = fkvKrutkoder.getBehandlingstema(journalpost.tema, journalpost.brevkode)
            val behandlingsType = fkvKrutkoder.getBehandlingstype(journalpost.tema, journalpost.brevkode)
            if (!(journalpost.behandlingstema == null || journalpost.behandlingstema!!.isEmpty())) {
                behandlingsTema = journalpost.behandlingstema
            }
            journalpost.behandlingstema = behandlingsTema
            journalpost.behandlingstype = behandlingsType
            log.info("Satt følgende verdier behandlingstema: '$behandlingsTema' og behandlingstype: '$behandlingsType' på journalpost ${journalpost.journalpostId}")
        } catch (e: Exception) {
            enrichedKafkaEvent.toFordeling = true
            enrichedKafkaEvent.isToManuell = true
            log.warn(
                "Klarte ikke finne behandlingsverdier for journalpost ${journalpost.journalpostId} med tema ${journalpost.tema} og skjema ${journalpost.brevkode}",
                e
            )
        }
    }

    private fun setupRefreshFellesKodeverkTimer() {
        val fellesKodeverkRefresher: TimerTask = object : TimerTask() {
            override fun run() {
                try {
                    fkvKrutkoder = fkvClient.fetchKrutKoder()
                } catch (e: Exception) {
                    log.error("Kunne ikke oppdatere KrutKoder fra FKV. Bruker gamle koder og prøver på nytt senere")
                }
            }
        }
        val executor = Executors.newSingleThreadScheduledExecutor()
        val refresherDelay = KRUT_KODER_REFRESH_INTERVAL_HOURS
        executor.scheduleAtFixedRate(fellesKodeverkRefresher, refresherDelay, refresherDelay, TimeUnit.HOURS)
    }
}
